### PR TITLE
TriggerTree and MostSpecificSelector

### LINKED ISSUE #1696: [PORT] TriggerTree and MostSpecificSelector classes
We need to port the `MostSpecificSelector` class from C# which has a dependency on TriggerTree.

Once the selector has been ported we need to update `AdaptiveDialog` to use this selector by default instead of `FirstSelector`.

### PR DESCRIPTION
**Pull request by chon219**
Fixes #1696 

## Description

Implemented TriggerTree and MostSpecificSelector.
Replaced default selector with MostSpecificSelector in adaptive dialogs.

## Specific Changes

  - Implemented TriggerTree and related components
  - Implemented MostSpecificSelector and made it as default selector
  - Changed interface of triggerSelector to return actions instead of action index
  - Made `evaluator` property in `Expression` public
  - Added negation in `ExpressionEvaluator`
  - Fixed a bug in `Expression` to make `deepEquals` work.

## Testing

- Ported tests for TriggerTree in adaptive-expressions.
- Enabled tests for MostSpecificSelector in adaptive-testing.

### CODE REVIEW

File: libraries/adaptive-expressions/src/triggerTrees/clause.ts [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+        }
+
+        builder.push(')');
+        if (this._ignored) {
+            builder.push(' ignored(');
+            builder.push(this._ignored.toString());
+            builder.push(')');
+        }
+
+        for (const key in this.anyBindings) {
```
Comment (joshgummersall) [inline libraries/adaptive-expressions/src/triggerTrees/clause.ts:82]: Not sure if our `tsconfig.json` currently supports [Object.entries](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries) but I tend to prefer this pattern:

```typescript
Object.entries(this.anyBindings).forEach(([key, value]) => builder.push(` ${ key }->${ value }`))
```

Many of these style suggestions (Array.prototype.forEach, Array.prototype.map, etc.) are not requirements, I just tend to prefer them. It's a little more straightforward and obvious than the subtlety of `for...of` vs. `for...in`.

File: libraries/adaptive-expressions/src/triggerTrees/clause.ts [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+
+    /**
+     * Compares the current `Clause` with another `Clause`.
+     * @param other The other `Clause` to compare.
+     * @param comparers A comparer, which is a dictionary of `PredicateComparer` with string keys.
+     * @returns A `RelationshipType` value between two `Clause` instances.
+     */
+    public relationship(other: Clause, comparers: { [name: string]: PredicateComparer }): RelationshipType {
+        let soFar: RelationshipType = RelationshipType.incomparable;
+        let shorter: Clause = this as Clause;
```
Comment (joshgummersall) [inline libraries/adaptive-expressions/src/triggerTrees/clause.ts:96]: I don't think you need both the `: Clause` and `as Clause` piece. Typescript will infer the type when using `as`.

File: libraries/adaptive-expressions/src/triggerTrees/clause.ts [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+        let shorterCount: number = shorter.children.length;
```
Comment (joshgummersall) [inline libraries/adaptive-expressions/src/triggerTrees/clause.ts:97]: The `: number` is probably unnecessary due to type inference. In general, we should leverage the Typescript compiler as much as possible. If we can omit type annotations that are inferred we should.

File: libraries/adaptive-expressions/src/triggerTrees/clause.ts [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+        return this.swap(soFar, swapped);
+    }
+
+    /**
+     * Determines whether the current `Clause` matches with another `Clause`.
+     * @param clause The other `Clause` instance to compare with.
+     * @param memory The scope for looking up variables.
+     * @returns A boolean value indicating whether the two clauses are matches.
+     */
+    public matches(clause: Clause, memory: any): boolean {
```
Comment (joshgummersall) [inline libraries/adaptive-expressions/src/triggerTrees/clause.ts:176]: Can we better annotate `memory` instead of using `any`?

File: libraries/adaptive-expressions/src/triggerTrees/clause.ts [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+            this._ignored = Expression.andExpression(...ignores);
+        }
+    }
+
+    private bindingRelationship(soFar: RelationshipType, shorterClause: Clause, longerClause: Clause): RelationshipType {
+        if (soFar === RelationshipType.equal) {
+            let swapped = false;
+            let shorter = shorterClause.anyBindings;
+            let longer = longerClause.anyBindings;
+            if (Object.entries(shorterClause.anyBindings).length > Object.entries(longerClause.anyBindings).length) {
```
Comment (joshgummersall) [inline libraries/adaptive-expressions/src/triggerTrees/clause.ts:216]: If `anyBindings` was a `Map` type the `.size` can be read in constant time rather than linear. It also might make iteration cleaner.

File: libraries/adaptive-expressions/src/triggerTrees/node.ts [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+    private _specializations: Node[] = [];
+
+    /**
+     * Intializes a new instance of the `Node` class.
+     * @param clause The logical conjunction this node represents.
+     * @param tree The trigger tree this node is found in.
+     * @param trigger The trigger to initialize this node.
+     */
+    public constructor(clause: Clause, tree: TriggerTree, trigger?: Trigger) {
+        this.clause = new Clause(clause);
```
Comment (joshgummersall) [inline libraries/adaptive-expressions/src/triggerTrees/node.ts:37]: Why reconstruct the `Clause`?

File: libraries/adaptive-expressions/src/triggerTrees/node.ts [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+    public relationship(other: Node): RelationshipType {
+        return this.clause.relationship(other.clause, this.tree.comparers);
+    }
+
+    /**
+     * Gets the most specific matches below this node.
+     * @param state Frame to evaluate against.
+     * @returns List of the most specific matches found.
+     */
+    public matches(state: any): Trigger[] {
```
Comment (joshgummersall) [inline libraries/adaptive-expressions/src/triggerTrees/node.ts:99]: Same issue WRT `state` being `any`.

File: libraries/adaptive-expressions/src/triggerTrees/node.ts [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+     * @param trigger The trigger to be removed.
+     * @returns Whether removing trigger operation is successful.
+     */
+    public removeTrigger(trigger: Trigger): boolean {
+        return this._removeTrigger(trigger, new Set<Node>());
+    }
+
+    private _addNode(triggerNode: Node, ops: Map<Node, Operation>): Operation {
+        let op = Operation.none;
+        if (!ops.has(this)) {
```
Comment (joshgummersall) [inline libraries/adaptive-expressions/src/triggerTrees/node.ts:125]: Early return is often preferable than deep nesting. In particular, this could be inverted and returned early to reduce the subsequent nesting.
```typescript
if (ops.has(this)) {
  return Operation.none;
}

...
```

File: libraries/adaptive-expressions/src/triggerTrees/node.ts [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+            const trigger = triggerNode.triggers[0];
+            const relationship = this.relationship(triggerNode);
+            switch (relationship) {
+                case RelationshipType.equal:
+                    // Ensure action is not already there
+                    let found = false;
+                    for (const existing of this._allTriggers) {
```
Comment (joshgummersall) [inline libraries/adaptive-expressions/src/triggerTrees/node.ts:132]: I like to use `Array.prototype.find` for these cases.

File: libraries/adaptive-expressions/src/triggerTrees/node.ts [hunk 10]
```diff
@@ -0,0 +1,362 @@
...
+
+            matched.set(this, found);
+        }
+
+        return found;
+    }
+
+    private _removeTrigger(trigger: Trigger, visited: Set<Node>): boolean {
+        let removed: boolean;
+        if (!visited.has(this)) {
```
Comment (joshgummersall) [inline libraries/adaptive-expressions/src/triggerTrees/node.ts:247]: Early return is preferred.

File: libraries/adaptive-expressions/src/triggerTrees/node.ts [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+            // Prevent visiting this node again
+            ops.set(this, op);
+        }
+
+        return op;
+    }
+
+    private _matches(state: any, matches: Set<Trigger>, matched: Map<Node, boolean>): boolean {
+        let found = matched.get(this);
+        if (!found) {
```
Comment (joshgummersall) [inline libraries/adaptive-expressions/src/triggerTrees/node.ts:218]: Early return is preferred.

File: libraries/adaptive-expressions/src/triggerTrees/trigger.ts [hunk 12]
```diff
@@ -0,0 +1,502 @@
...
+
+            if (changed) {
+                newExpr = new Expression(undefined, expression.evaluator, ...children);
+            }
+        }
+
+        return { expression: newExpr, changed };
+    }
+
+    private removeDuplicates(): void {
```
Comment (joshgummersall) [inline libraries/adaptive-expressions/src/triggerTrees/trigger.ts:483]: It seems like use of underscore prefixes is not consistent - what is the convention or rule that we try to follow?

File: libraries/adaptive-expressions/tests/triggerTree.test.js [hunk 13]
```diff
@@ -0,0 +1,656 @@
...
+                bindings[key] = info.bindings[key];
+            }
+        }
+        return bindings;
+    }
+
+    randomString(length) {
+        let result = '';
+        for (let i = 0; i < length; i++) {
+            const chars = 'abcdefghijklmnopqrstuvwxyz';
```
Comment (joshgummersall) [inline libraries/adaptive-expressions/tests/triggerTree.test.js:309]: There's no need to create `chars` inside the for-loop.

File: libraries/adaptive-expressions/tests/triggerTree.test.js [hunk 14]
```diff
@@ -0,0 +1,656 @@
...
+        return result;
+    }
+
+    randomChoice(choices) {
+        const selection = Math.floor(Math.random() * choices.length);
+        return choices[selection];
+    }
+
+    randomWeighted(weights) {
+        let totalWeight = 0;
```
Comment (joshgummersall) [inline libraries/adaptive-expressions/tests/triggerTree.test.js:322]: Up to you here, but `Array.prototype.reduce` is handy for this.
```typescript
const totalWeight = weights.reduce((sum, weight) => sum + weight, 0);
```

File: libraries/adaptive-expressions/tests/triggerTree.test.js [hunk 15]
```diff
@@ -0,0 +1,656 @@
...
+
+    adjustIntValue(value, type) {
+        let result = value;
+        const epsilon = 1;
+        switch (type) {
+            case ExpressionType.LessThan:
+                result += epsilon;
+                break;
+            case ExpressionType.NotEqual:
+                result += epsilon;
```
Comment (joshgummersall) [inline libraries/adaptive-expressions/tests/triggerTree.test.js:350]: Collapse this case with the one above.

File: libraries/adaptive-expressions/tests/triggerTree.test.js [hunk 16]
```diff
@@ -0,0 +1,656 @@
...
+                break;
+        }
+
+        return isNot ? undefined : { type: comparison.type, value };
+    }
+
+    variablesByType(bindings) {
+        const result = {};
+        for (const key in bindings) {
+            const value = bindings[key].value;
```
Comment (joshgummersall) [inline libraries/adaptive-expressions/tests/triggerTree.test.js:451]: Destructured assignment is handy here: `const { value } = bindings[key];`

**Review by joshgummersall — COMMENTED** [review]
Lots of suggestions here and there and a few questions. I can't really comment on all the functionality as I'm not familiar enough with the libraries just yet.

File: libraries/adaptive-expressions/src/triggerTrees/quantifier.ts [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+ * Quantifier for allowing runtime expansion of expressions.
+ */
+export class Quantifier {
+    /**
+     * Initializes a new instance of the `Quantifier` class.
+     * @param variable Name of variable to replace.
+     * @param type Type of quantifier.
+     * @param bindings Possible bindings for variable.
+     */
+    public constructor(variable: string, type: QuantifierType, bindings: string[]) {
```
[the remark this answers is not in the document]
Comment (chon219) [reply libraries/adaptive-expressions/src/triggerTrees/quantifier.ts:34]: Cool!

File: libraries/adaptive-expressions/src/triggerTrees/trigger.ts [hunk 18]
```diff
@@ -0,0 +1,502 @@
...
+    private _clauses: Clause[];
+
+    /**
+     * Intializes a new instance of the `Trigger` class.
+     * @param tree Trigger tree that contains this trigger.
+     * @param expression Expression for when the trigger action is possible.
+     * @param action Action to take when a trigger matches.
+     * @param quantifiers Quantifiers to dynamically expand the expression.
+     */
+    public constructor(tree: TriggerTree, expression: Expression, action: any, ...quantifiers: Quantifier[]) {
```
[the remark this answers is not in the document]
Comment (chon219) [reply libraries/adaptive-expressions/src/triggerTrees/trigger.ts:86]: Actually the `action` can be anything, it can be a string or an object or something else, it is just a container for real "actions". e.g. In trigger tree tests, we use strings as actions. In MostSpecificSelector, we use `OnCondition` objects as actions.
When I was porting this class from C#, the type of action was `object` because anything can be object in c#. So I have to leave it `any` here.

File: libraries/adaptive-expressions/src/triggerTrees/trigger.ts [hunk 19]
```diff
@@ -0,0 +1,502 @@
...
+        return this._clauses;
+    }
+
+    /**
+     * Determines the relationship between current instance and another `Trigger` instance.
+     * @param other The other Trigger instance.
+     * @param comparers The comparer dictionary.
+     * @returns A `RelationshipType` value.
+     */
+    public relationship(other: Trigger, comparers: { [name: string]: PredicateComparer }): RelationshipType {
```
[the remark this answers is not in the document]
Comment (chon219) [reply libraries/adaptive-expressions/src/triggerTrees/trigger.ts:129]: You are right, it should be extracted.

File: libraries/adaptive-expressions/src/triggerTrees/trigger.ts [hunk 20]
```diff
@@ -0,0 +1,502 @@
...
+        return result;
+    }
+
+    /**
+     * Determines whether there is a member in the current `Clause` that matches the nodeClause parameter.
+     * @param nodeClause The other Clause instance to match.
+     * @param state The scope for looking up variables.
+     * @returns A boolean value inidicating whether there is a member matches.
+     */
+    public matches(nodeClause: Clause, state: any): boolean {
```
[the remark this answers is not in the document]
Comment (chon219) [reply libraries/adaptive-expressions/src/triggerTrees/trigger.ts:160]: I would use `MemoryInterface | any` instead to align with what we did in `Expression.tryEvaluate()`.

File: libraries/adaptive-expressions/src/triggerTrees/trigger.ts [hunk 21]
```diff
@@ -0,0 +1,502 @@
...
+                }
+                return clauses;
+            case ExpressionType.Optional:
+                return [
+                    new Clause(),
+                    ...this.generateClauses(expression.children[0])
+                ];
+            default:
+                // True becomes empty expression and false drops clause
+                if (expression instanceof Constant && typeof (expression.value) === 'boolean') {
```
[the remark this answers is not in the document]
Comment (chon219) [reply libraries/adaptive-expressions/src/triggerTrees/trigger.ts:283]: if `expression.value` is false, we will return an empty list.
I would use the following code instead:
```
return expression.value ? [new Clause()] : []
```

File: libraries/adaptive-expressions/src/triggerTrees/node.ts (hunk 6 above)
Comment (chon219) [reply libraries/adaptive-expressions/src/triggerTrees/node.ts:37]: I was porting the c# implementation, it is supposed to deep clone the clause here with `newClause = new Clause(fromClause)`.

File: libraries/adaptive-expressions/src/triggerTrees/trigger.ts (hunk 12 above)
Comment (chon219) [reply libraries/adaptive-expressions/src/triggerTrees/trigger.ts:483]: I would use underscores for all private methods.

File: libraries/adaptive-expressions/tests/triggerTree.test.js (hunk 14 above)
Comment (chon219) [reply libraries/adaptive-expressions/tests/triggerTree.test.js:322]: Thanks for reminding this.

File: libraries/adaptive-expressions/src/triggerTrees/clause.ts (hunk 5 above)
Comment (chon219) [reply libraries/adaptive-expressions/src/triggerTrees/clause.ts]: Changed `anyBindings` to `Map`.

File: libraries/adaptive-expressions/src/triggerTrees/clause.ts (hunk 4 above)
Comment (chon219) [reply libraries/adaptive-expressions/src/triggerTrees/clause.ts]: Changed type of `memory` to `MemoryInterface | any` to match the `tryEvaluate()` method.

File: libraries/adaptive-expressions/src/triggerTrees/node.ts (hunk 7 above)
Comment (chon219) [reply libraries/adaptive-expressions/src/triggerTrees/node.ts]: Changed type of `state` to `MemoryInterface | any` to match the `tryEvaluate()` method.

**Review by joshgummersall — COMMENTED** [review]
The code all looks good to me. I would defer to someone with more specific knowledge of the dotnet TriggerTree implementation for final sign off.

File: libraries/adaptive-expressions/src/expression.ts [hunk 22]
```diff
@@ -41,7 +41,7 @@ export class Expression {
      */
     public children: Expression[];
 
-    protected readonly evaluator: ExpressionEvaluator;
+    public readonly evaluator: ExpressionEvaluator;
```
Comment (chrimc62) [inline libraries/adaptive-expressions/src/expression.ts:45]: If you need to make this public, should add documentation.

File: libraries/botbuilder-dialogs-adaptive/src/actions/dynamicBeginDialog.ts [hunk 23]
```diff
@@ -6,12 +6,18 @@
...
 import { DialogContext } from 'botbuilder-dialogs';
 import { BeginDialog } from './beginDialog';
 
 export class DynamicBeginDialog extends BeginDialog {
-    protected bindOptions(dc: DialogContext, options: object): object {
-        // use overflow properties of deserialized object instead of the passed in option.
-        return super.bindOptions(dc, Object.assign({}, this));
+    protected bindOptions(dc: DialogContext, _options: object): object {
+        const options = {};
+        for (const key of Object.getOwnPropertyNames(this)) {
```
Comment (chrimc62) [inline libraries/botbuilder-dialogs-adaptive/src/actions/dynamicBeginDialog.ts:16]: This is not really related to trigger trees, right?

**Review by chrimc62 — APPROVED** [review]
:shipit:

File: libraries/botbuilder-dialogs-adaptive/src/actions/dynamicBeginDialog.ts (hunk 23 above)
Comment (chon219) [reply libraries/botbuilder-dialogs-adaptive/src/actions/dynamicBeginDialog.ts:16]: No, but if I didn't make this change, the test will fail.
